[DCP] overwrites existing checkpoint by default

LucasLLC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125877

Checks for existing checkpoints and overwrites, based on an `overwrite` flag

Differential Revision: [D57186174](https://our.internmc.facebook.com/intern/diff/D57186174/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

LucasLLC committed:
[DCP] overwrites existing checkpoint by default

Checks for existing checkpoints and overwrites, based on an `overwrite` flag

Differential Revision: [D57186174](https://our.internmc.facebook.com/intern/diff/D57186174/)

[ghstack-poisoned]

wz337 commented on test/distributed/checkpoint/test_fsspec.py:
```diff
@@ -176,6 +177,25 @@ def opt_at(opt, idx):
+        self.assertTrue(torch.allclose(sd["random"], t2))
+
+        dcp.save(
+            {"random": t2}, storage_writer=FsspecWriter(self.temp_dir, overwrite=True)
```
Should this be overwrite=False when exception is raised, As the default is overwrite=True.

LucasLLC replied:
yes, and the format for assert raises is wrong as well, I'll have to take a look

LucasLLC replied:
fixed, thanks @wz337 !

LucasLLC committed:
Update on "[DCP] overwrites existing checkpoint by default"


Checks for existing checkpoints and overwrites, based on an `overwrite` flag

Differential Revision: [D57186174](https://our.internmc.facebook.com/intern/diff/D57186174/)

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

fegin reviewed (commented):
It's good to have the overwritten feature. But shouldn't this be an opt-in feature? Just like the `cp` linux command will explicitly ask if users would want to overwrite the existing checkpointing. If we do this implicitly, would it cause some surprises to the users?

LucasLLC commented:
> It's good to have the overwritten feature. But shouldn't this be an opt-in feature? Just like the `cp` linux command will explicitly ask if users would want to overwrite the existing checkpointing. If we do this implicitly, would it cause some surprises to the users?

good point, will update

LucasLLC committed:
Update on "[DCP] overwrites existing checkpoint by default"


Checks for existing checkpoints and overwrites, based on an `overwrite` flag

Differential Revision: [D57186174](https://our.internmc.facebook.com/intern/diff/D57186174/)

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

LucasLLC commented:
Will fix errors before merging

LucasLLC committed:
Update on "[DCP] overwrites existing checkpoint by default"


Checks for existing checkpoints and overwrites, based on an `overwrite` flag

Differential Revision: [D57186174](https://our.internmc.facebook.com/intern/diff/D57186174/)

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

LucasLLC commented:
> Will fix errors before merging



> > It's good to have the overwritten feature. But shouldn't this be an opt-in feature? Just like the `cp` linux command will explicitly ask if users would want to overwrite the existing checkpointing. If we do this implicitly, would it cause some surprises to the users?
> 
> good point, will update

We decided actually that setting `overwrite=True` as the default is safer due to BC considerations, since it turns out depending on the implementation (fsspec vs. filesystem), overwrite may or may not error when attempting to overwrite, but in the case of this PR will always raise if `overwrite=False`

I added a warning when we detect a checkpoint, and will change `overwrite=False` as the default after 2.5

LucasLLC commented:
@pytorchbot merge
